### DISCUSSION TITLE
Prevent direct access to tables via REST but allow from functions

### DISCUSSION THREAD

**Discussion by Sesa1988** [supabase/supabase#7574] [Questions] [answered]
# Feature request

## Is your feature request related to a problem? Please describe.

I want to limit the amount a user can create based on his purchase status. 
I have to call another REST interface to get his status and for this I need to use PostgreSQL functions or Edge Functions since RSL Rules can't do http calls as far as I know.
 
My issue is someone could just call the PostgreSQL API directly and create as much as he wants.

I could create a PostgreSQL function with "superuser" rights but that feels like a workaround. I don't know if someone could abuse this even if I check for the passed user id there. We are getting the user id from the provided token right?

## Describe the solution you'd like

A setting for each table that prevents direct calls but allow calls from PostgreSQL functions or Edge Functions with the same configured RSL Rules applied.

There is no other way currently right? Is there a security concern with that PostgreSQL function and super user solution described above?

**Comment by w3b6x9** [comment]
@Sesa1988 if your users are `authenticated` then you can create RLS policies where you can query based on the `authenticated` user claims and check if that user has authorization to create.

Also, you mention RSL Rules a few times. What are RSL Rules?

**Comment by steve-chavez** [comment, marked as answer]
> My issue is someone could just call the PostgreSQL API directly and create as much as he wants.
> Prevent direct access to tables via REST but allow from functions

You can create a `private` schema and have your tables there and then only create functions in the `public` schema(here you can refer to tables in the `private` schema). You can see more details about this pattern [here](https://postgrest.org/en/stable/schema_structure.html#schema-isolation).

**Reply by parkernilson** [reply]
I would like to do something similar, however I need users to be able to SELECT from the tables in private, but I only want them to be able to modify them through functions. Is this possible?

**Comment by Sesa1988** [comment]
> > My issue is someone could just call the PostgreSQL API directly and create as much as he wants.
> > Prevent direct access to tables via REST but allow from functions
> 
> You can create a `private` schema and have your tables there and then only create functions in the `public` schema(here you can refer to tables in the `private` schema). You can see more details about this pattern [here](https://postgrest.org/en/stable/schema_structure.html#schema-isolation).

Thats exactly what I need, thx.

**Comment by Sesa1988** [comment]
> > My issue is someone could just call the PostgreSQL API directly and create as much as he wants.
> > Prevent direct access to tables via REST but allow from functions
> 
> You can create a `private` schema and have your tables there and then only create functions in the `public` schema(here you can refer to tables in the `private` schema). You can see more details about this pattern [here](https://postgrest.org/en/stable/schema_structure.html#schema-isolation).

I have to run my function with "security definer" because else you cant access any private tables from a public function. 

Im not sure if this is a security concern. 

It looks like you can set a flag to enable row security even the function runs by privilage of the creator of the function. This way I should have a function with creator privilage but row security based on the passed user token to the function (by the framework) ?

**Reply by steve-chavez** [reply]
> I have to run my function with "security definer" because else you cant access any private tables from a public function.
> Im not sure if this is a security concern.

Note that the security definer role doesn't have to be a superuser. It can be another newly created pg role, say "rpc_role", that has explicitly assigned privileges on the private tables you wish to grant access to in the public function.

And yes, you can even have RLS working through this security definer role.

**Reply by steve-chavez** [reply]
Basically:

```sql
create role rpc_role;

grant select, insert on private.tbl to rpc_role;
...

begin;
set local role rpc_role;
create or replace function pubic.my_func()... 
commit;
```

**Comment by vduseev** [comment]
I've tried to achieve the same and spent a few hours digging in the wrong direction. Here is what have worked and what won't.

*Extremely important detail: the queries below are executed by `postgres` user, which is not a real superuser in Supabase. It is [explicitly demoted](https://github.com/supabase/postgres/blob/develop/migrations/db/migrations/10000000000000_demote-postgres.sql) during database creation. Only `supabase_admin` is a true superuser. All migrations are executed by `postgres` user, however. That has a very important implication: you can't grant access on, say, `auth` schema to a newly created role, because `postgres` user has no rights to do so.*

**Working solution**:

```sql
-- Create new private schema.
-- It does not necessarily need to be called private. It can be whatever. The important
-- thing is that `anon` and `authenticated` roles have no access to it.
CREATE SCHEMA IF NOT EXISTS private;

-- If you had tables that you want to move to new private schema, you can move them
ALTER TABLE my_table SET SCHEMA private;

-- Or create a new table inside the private schema
CREATE TABLE private.my_profiles (
  user_id UUID NOT NULL REFERENCES auth.users (id) ON DELETE CASCADE,
  nickname TEXT
);

-- Now you need to create a function that will be owned by `postgres` user and, when invoked,
-- will be executed with `postgres` permissions as well. Adding SECURITY DEFINER at the end
-- of the function definition helps us achieve that last part.
CREATE OR REPLACE FUNCTION public.my_function (
  p_user_id UUID
) RETURNS TABLE (
  user_id UUID,
  nickname TEXT
) AS $$
BEGIN
  -- Only allow users to request their own profile.
  -- Notice how we can use functions from auth schema here, because the function we are creating
  -- is owned by postgres user, which has access to auth schema granted to it by supabase_admin
  IF p_user_id != auth.uid() THEN
    RAISE EXCEPTION 'permission denied: you cannot get info for this p_user_id';
  END IF;

  RETURN QUERY
  SELECT
    t.user_id,
    t.nickname
  FROM private.my_profiles t
  WHERE t.user_id = p_user_id;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

Now if you create a pgTAP test to verify if that works it will demonstrate that everything works as intended.

```sql
BEGIN;
SELECT no_plan();

set local role authenticated;
-- I have added the user below to my database in advance, using seed.sql
set local "request.jwt.claim.sub" to '17d19f43-b764-4fa0-b7ca-49c5a2d1c33c';

-- Users can only query their own data
SELECT throws_like(
  $$SELECT * FROM my_function('fbae6049-ddb7-41ea-8a02-02418d81be24')$$,
  'permission denied%',
  'p_user_id must be same as UID of authenticated user'
);

-- Users cannot read from private tables
SELECT throws_like(
  'SELECT * FROM private.my_profiles',
  '%permission denied%',
  'User cannot read directly from private.* tables'
);

-- Users can query their own data just fine
SELECT is(
  (SELECT COUNT(*)::INTEGER FROM my_function('17d19f43-b764-4fa0-b7ca-49c5a2d1c33c')),
  1,
  'my_function returns 1 record that matches current user'
);
```

**What's not going to fully work**

Creating a separate role like `rpc_role` won't work fully. You can create it alright and even grant it access to tables in the `private` schema. But you can't grant this role access to `auth`, `storage`, and other schemas that are owned by `supabase_admin`. Which means your functions won't have access to `auth.uid()` and will throw errors like:

```shell
ERROR:  permission denied for schema auth
```

This means that creating `rpc_role` is a fine approach as long as you don't need access to `supabase_admin` owned schemas inside them, such as `auth`, `storage`, etc.

I've tried performing `supabase db reset` with `--db-url` connection specifying `supabase_admin` user, since it has the same password as `postgres` by default. Unfortunately, that does not have any effect since `supabase db reset` command switches to `postgres` user anyway and still can't grant privileges on `auth` schema to non-postgres roles.

**Reply by bombillazo** [reply]
Not having access to storage/auth kills this pattern... 😞

**Comment by Braffolk** [comment]
Late to the party, but I found a working solution. Honestly would have expected this feature to be built in, to be able to limit data api access to only edge functions.

tl;dr add a custom claim to the JWT in edge functions that RLS will require (create new JWT token, copy important bits from old token). This means users can still directly attempt to send requests to the Data API, but they will always get an empty response. Now if an edge function makes the same request with the new user token, RLS will apply normally and the request will process as expected.

This was written during late hours, so might not be perfectly optimal yet.

```Typescript

const SUPABASE_URL = Deno.env.get('SUPABASE_URL')!;
const SUPABASE_ANON_KEY = Deno.env.get('SUPABASE_ANON_KEY')!;
const JWT_SECRET = Deno.env.get('JWT_SECRET')!;

async function mintToken(cognitoToken: string | undefined): Promise<string> {
    const now = Math.floor(Date.now() / 1000);
    
    // Default claims for anonymous users
    const payload: Record<string, any> = {
        aud: "anon",
        exp: now + 5 * 60, 
        iat: now,
        iss: "supabase",
        sub: crypto.randomUUID(),
        role: "anon",
        edge_authorized: true // Our custom claim
    };

    // If cognito token exists, extract user identity
    if (cognitoToken) {
        try {
            // Decode token. Might be good to double verify it here.
            const parts = cognitoToken.split('.');
            if (parts.length === 3) {
                const payloadStr = atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'));
                const decoded = JSON.parse(payloadStr);
                
                // Copy user identity details
                payload.sub = decoded.sub || payload.sub;
                payload.email = decoded.email;
               // extract the relevant data for new token
           }
        } catch (error) {
            console.error("Failed to decode Cognito token:", error);
            // Keep anonymous identity if decode fails
        }
    }
    
    // Create key from JWT secret
    const keyData = new TextEncoder().encode(JWT_SECRET);
    const key = await crypto.subtle.importKey(
        "raw",
        keyData,
        { name: "HMAC", hash: "SHA-256" },
        false,
        ["sign"]
    );

    return await create(
        { alg: "HS256", typ: "JWT" },
        payload,
        key
    );
}

export async function initClient(authHeader: string | null): Promise<SupabaseClient<Database>> {
    let headers: Record<string, string> = {};
    if (authHeader) {
        const cognitoToken = authHeader.split(' ')[1];
    
        const enhancedToken = await mintToken(cognitoToken);
        headers = {
            Authorization: `Bearer ${enhancedToken}`
        };
    } else {
        // Create anonymous token with edge_authorized flag
        const anonToken = await mintToken(undefined);
        headers = {
            Authorization: `Bearer ${anonToken}`
        };
    }

    const supabase = createClient<Database>(
        SUPABASE_URL,
        SUPABASE_ANON_KEY,
        {
            global: {
                headers: headers
            }
        }
    );

    return supabase;
}
```

It's critical for me to handle both anon and logged in requests, hence why I allow both missing and in-place tokens. For others use case, this might not be necessary.

Now for the RLS part I have set up a helper SQL function to verify the custom claim:

```SQL
CREATE OR REPLACE FUNCTION is_request_from_edge_function() RETURNS boolean AS $$
BEGIN
  RETURN coalesce(
    current_setting('request.jwt.claims', true)::json->>'edge_authorized',
    'false'
  ) = 'true';
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

To use the function I define my RLS as follows:
```SQL
CREATE POLICY "survey_question_i18n_edge_function_rls" ON "public"."survey_question_i18n"
  AS PERMISSIVE FOR SELECT TO authenticated
  USING (is_request_from_edge_function());
```

```SQL
CREATE POLICY "dashboard_access_groups_rls"
    ON "public"."dashboard"
    AS PERMISSIVE
    FOR SELECT
    TO authenticated
    USING (
        is_request_from_edge_function() AND (
            cardinality(access_groups) = 0
            OR (cache_cognito_groups() && access_groups)
        )
    );
```

Obviously this can and should be optimized further, potentially caching the tokens in memory instead of "minting" them on every request, but works well for a starting point.

Cognito is not relevant here, will work with any provider. It's just what I use.
